keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The service controller in the Kubernetes cloud controller manager (KCCM)
+configures the load balancer and its corresponding health check (HC) following
+service related events. The configured health check is then used by the load
+balancer as to determine which instances are candidates for traffic load
+balancing. For certain cloud providers (GCP): the KCCM configures the HC to
+target the Kubernetes service proxy for this information. In this KEP we will
+focus on Kube-proxy, since it's the only service proxy under the responsibility
+of the Kubernetes project.
+
+We can define two classes of services for what concerns the HC:
+
+- A) services which have `externalTrafficPolicy: Cluster`
+- B) services which have `externalTrafficPolicy: Local`
+
+For class A) services: Kube-proxy currently returns an answer following its
+`healthz` state. For class B) services: Kube-proxy only reports if the service
+for which the load balancer was created for, has a `Ready` endpoint running on
+the node.
+
+This KEP proposes two changes:
+
+1. That Kube-proxy's includes the `healthz` state AND its current answer w.r.t
+  the local endpoints, when it answers to the HC for class B) services.
+
+2. That Kube-proxy enables connection draining for terminating Nodes. This is to
+  be done by Kube-proxy by inspecting a field on the Node object which indicates
+  that the "node is terminating/deleting" and when seen, starts failing the HC.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+We need to agree on this field. `deletionTimestamp` is the perfect indicator for
+this, but....the Node object does not have a finalizer, so that wouldn't help.
```

Comment:
...and at least some cases, like CA, the underlying VM is deleted before the Node.

Reply:
Right, I have too much context which I didn't spell out. I meant to say: "if the Node object had finalizer(s) and controllers triggered the delete of the Node object and subsequently removed their finalizer(s)" That would have been a clear indicator for all parties

Reply:
That's a very large "if" :)

Reply:
what is CA

>  the underlying VM is deleted before the Node.

if the VM is deleted, kube-proxy no longer exists XD

Reply:
CA is "Cluster Autoscaler"

My point is that if we define the trigger for "accepting" -> "drainign" to be `deletionTimestamp` it won't work because the thing we want to avoid (abrupt termination without drain) has already happened.